LSM: ManifestLevel fuzzer

kprotty opened the pull request:
Prototype of a fuzzer which simulates ManifestLevel usage during Manifest under various interleavings. Needs a bit more optimizations and verification, table visibility, and maybe better FuzzOp generation.

## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance.

kprotty committed:
lsm: fuzz_manifest_level prototype

jorangreef commented on src/lsm/manifest_level_fuzz.zig:
```diff
@@ -0,0 +1,426 @@
+//! - Tables with updated snapshot_max become invisible to future snapshots from the iterator.
+//! - Tables visible to the current/future snapshot can be removed.
+//! - Tables invisible to the current/future/past snapshots can be removed.
+//! - Tables can be inserted, updated, and removed in any order.
```
```
+ All tables can be updated.
+ All tables can be removed.
```

jorangreef replied:
We're still missing `Strategy` here, a short sentence or paragraph to tell the story of how the fuzzer maximizes coverage-to-test-code ratio.

jorangreef commented on src/lsm/manifest_level_fuzz.zig:
```diff
@@ -0,0 +1,426 @@
+//! Fuzz ManifestLevel insert_table()/set_snapshot_max()/remove_table*()/iterator().
```
We should indicate whether all public methods are covered here, so that the reader doesn't have to work this out.

For example, `All public methods are covered.`

jorangreef commented on src/lsm/manifest_level_fuzz.zig:
```diff
@@ -0,0 +1,426 @@
+                    .{},
+                );
+
+                // Can't be equal as the tables may not overlap
```
Comments are sentences and should always end with punctuation.

For example, either `.` if they refer clearly to the next line in isolation, or `:` if they refer to multiple lines beneath.

jorangreef commented on src/lsm/manifest_level_fuzz.zig:
```diff
@@ -0,0 +1,426 @@
+
+            return .{
+                .checksum = env.random.int(u128),
+                // zero addresses are used to indicate the table being removed.
```
```suggestion
                // Zero addresses are used to indicate the table being removed.
```

Proper capitalization tells the reader we're sharp about the little things.

jorangreef reviewed (commented):
Nice work @kprotty.

Commit:
lsm: improve ManifestLevel fuzzer docs & op generation

The state of the ManifestLevel should now be accurately tracked during
FuzzOp generation. This ensures invalid FuzzOps aren't generated and
allows the Environment to assert on expected transformations. The
overall strategy and invariants for the doc comment were also fleshed out.

Commit:
lsm: ManifestLevel fuzzer comment punctuation

Commit:
lsm: zig fmt

kprotty committed:
lsm: use binary_search for ManifestLevel FuzzOp.update_tables

Commit:
lsm: bump level fuzz count & remove with O(log n)

batiati commented on src/lsm/manifest_level_fuzz.zig:
```diff
@@ -0,0 +1,497 @@
+//!
+//! - Tables visible to both snapshot_latest and the current snapshot can be removed.
+//! - Tables invisible to snapshot_latest but still the current snapshot cannot be removed.
+//!     - The current snapshot must be bumped which puts them in the next category:
```
Not a big deal, but I felt like this `tab` made the comment harder to understand than it actually is. At least for me, I had to read 3 times to be sure that I wasn't missing the relationship between those 3 last invariants.

I guess just a plain list would be clear enough:
```suggestion
//! - The current snapshot must be bumped which puts them in the next category:
```

batiati commented on src/lsm/manifest_level_fuzz.zig:
```diff
@@ -0,0 +1,497 @@
+
+                // Decide if all tables visible ot snapshot_latest should be removed.
+                var amount = ctx.random.intRangeAtMostBiased(usize, 1, visible_latest);
+                if (ctx.random.boolean()) amount = visible_latest;
```
I got confused by this second assignment on `amount`.

If the intention is to have 50% chance for `visible_latest` and 50% chance for between `1` and `visible_latest`,  I'd suggest this instead:

```zig
 // Decide if all tables visible ot snapshot_latest should be removed.
var amount = if (ctx.random.boolean()) visible_latest else ctx.random.intRangeAtMostBiased(usize, 1, visible_latest);
```

kprotty replied:
That was the original plan, but it didn't fit in the 100 column limit. What about a full if statement?
```zig
var amount = if (ctx.random.boolean())
    visible_latest
else
    ctx.random.intRangeAtMostBiased(usize, 1, visible_latest);
```

batiati replied:
Works great!

batiati commented on src/lsm/manifest_level_fuzz.zig:
```diff
@@ -0,0 +1,497 @@
+                    return ctx.next(.remove_invisible);
+                }
+
+                // Decide if all tables visible ot snapshot_latest should be removed.
```
```suggestion
                // Decide if all tables visible at snapshot_latest should be removed.
```

batiati reviewed (commented):
Neat work @kprotty!

kprotty committed:
Apply suggestions from code review

Co-authored-by: Rafael Batiati <rbatiati@gmail.com>

kprotty committed:
lsm: use inline-if in manifest_level_fuzz
